packages/commons/src/interop-token/models.ts

```diff
@@ -3,134 +3,255 @@ import {
+export const InteropJwtApiM2MAdminPayload = InteropJwtApiCommonPayload.merge(
+  z.object({
+    role: z.literal(systemRole.M2M_ADMIN_ROLE),
+    adminId: z.string().uuid(),
+    // ^ ID of the admin user associated with the client
+  })
+);
+export type InteropJwtApiM2MAdminPayload = z.infer<
+  typeof InteropJwtApiM2MAdminPayload
+>;
 
+// ==========================================
+//   Interop API Token
+// ==========================================
 export type InteropJwtApiPayload =
   | InteropJwtApiM2MAdminPayload
   | InteropJwtApiM2MPayload;
 
+/* ========================================== 
+    Interop API Token 
+  ========================================== */
+
 export type InteropApiToken = {
   header: InteropJwtHeader;
   payload: InteropJwtApiPayload;
   serialized: string;
 };
 
-/* ========================================== 
-    Interop INTERNAL Token 
-  ========================================== */
-export type InteropJwtInternalPayload = InteropJwtCommonPayload & {
-  sub: string;
-  role: Extract<SystemRole, "internal">;
-};
+// ==========================================
+//   Interop INTERNAL Token
+// ==========================================
+export const InternalJwtAuthPayload = InteropJwtCommonPayload.merge(
+  z.object({
+    role: z.literal(systemRole.INTERNAL_ROLE),
+    sub: z.string(),
+  })
+);
+export type InternalJwtAuthPayload = z.infer<typeof InternalJwtAuthPayload>;
+
+// ==========================================
+//  Interop MAINTENANCE Token
+// ==========================================
+export const InteropJwtMaintenancePayload = InteropJwtCommonPayload.merge(
+  z.object({
+    role: z.literal(systemRole.MAINTENANCE_ROLE),
+    sub: z.string(),
+  })
+);
+export type InteropJwtMaintenancePayload = z.infer<
+  typeof InteropJwtMaintenancePayload
+>;
+
+// ==========================================
+//   Interop UI Token
+// ==========================================
+export const InteropJwtUIPayload = InteropJwtCommonPayload.merge(
+  z.object({
+    // setting role to z.undefined() to make the discriminated union work.
+    // z.discriminatedUnion performs better than z.union and gives more meaningful parsing errors.
+    role: z.undefined(),
+    "user-roles": CommaSeparatedStringToArray(UserRole),
+    uid: z.string().uuid(),
+    organizationId: z.string().uuid(),
+    selfcareId: z.string().uuid(),
+    organization: z.object({
+      id: z.string().uuid(),
+      name: z.string(),
+      roles: z.array(
+        z.object({
+          partyRole: z.string().nullish(),
+          role: UserRole,
+        })
+      ),
+      fiscal_code: z.string().nullish(),
+      ipaCode: z.string().nullish(),
+    }),
+    externalId: z.object({
+      origin: z.string(),
+      value: z.string(),
+    }),
+    name: z.string().nullish(),
+    family_name: z.string().nullish(),
+    email: z.string().nullish(),
+  })
+);
+export type InteropJwtUIPayload = z.infer<typeof InteropJwtUIPayload>;
+
```

Comment:
Same here, in the previous sections, you also defined the token, not only the payload - see comment/question above